Add support ignore files

gucong3000 opened the pull request:
- [x] Added: `--stdin-filename` option to CLI. same as [stylelint CLI](https://github.com/stylelint/stylelint/blob/master/src/cli.js#L75)
- [x] Added: `--ignore-path` option to CLI. same as [stylelint CLI](https://github.com/stylelint/stylelint/blob/master/src/cli.js#L63)
- [x] Added: `ignorePath` option to Node.js API. same as [stylelint API](http://stylelint.io/user-guide/node-api/#ignorepath)
- [x] Added: `ignoreFiles` option to configuration file. same as [stylelint configuration](http://stylelint.io/user-guide/configuration/#ignorefiles)

morishitter commented on bin/cli.js:
```diff
@@ -38,14 +40,16 @@ if (argv.h) {
+  console.log('  -b, --config-basedir   Path to the directory that relative paths defining "extends"')
+  console.log('  -v, --version          Output the version number')
+  console.log('  -h, --help             Output usage information')
+  console.log('  -i, --ignore-path      Path to a file containing patterns that describe files to ignore.')
```
is this the path of `.stylelintignore` file ?

morishitter replied:
i think, `--config-path` is more suitable

gucong3000 replied:
A path to a file containing patterns describing files to ignore. The path can be absolute or relative to process.cwd(). By default, stylelint looks for .stylelintignore in process.cwd().
See: http://stylelint.io/user-guide/node-api/#ignorepath

morishitter replied:
Thanks, I understood what `--ignore-path` option means, but, I'm sorry, I can't understand what `--ignore-disable` option means clearly.

gucong3000 replied:
I just removed `--ignore-disable` option, because it's so much different from `stylelint`.

morishitter commented on bin/cli.js:
```diff
@@ -38,14 +40,16 @@ if (argv.h) {
+  console.log('  -h, --help             Output usage information')
+  console.log('  -i, --ignore-path      Path to a file containing patterns that describe files to ignore.')
+  console.log('  -id --ignore-disables  Ignore disables')
+  console.log('  --stdin-filename       A filename to assign stdin input.')
```
it seems strange name for me, why "filename"? this is just stdin.

gucong3000 replied:
This can be useful, for example, when making a text editor plugin that passes in code directly but needs to still use the configuration's ignoreFiles functionality to possibly ignore that code.

See:
https://github.com/stylelint/stylelint/blob/master/src/cli.js#L155
http://stylelint.io/user-guide/node-api/#codefilename

gucong3000 replied:
[sublime-stylefmt](https://github.com/dmnsgn/sublime-stylefmt) Have a [stylefmt caller](https://github.com/dmnsgn/sublime-stylefmt/blob/master/stylefmt.js).

I think `sublime-stylefmt` should just call `cli.js` of `stylefmt`, not creat other caller, so we need this API

`sublime-stylefmt` should just call `stylefmt --stdin-filename /path/to/a/change/unsaved/file`

If user of SublimeText call stylefmt before save file, we need pass file path and code, Not a single one can be omitted

morishitter replied:
I see, thanks :)

morishitter commented on index.js:
```diff
@@ -10,16 +10,18 @@ const formatSassVariables = require('./lib/formatSassVariables')
-      formatRules(root, params)
-      formatSassVariables(root, params)
-      return root
+  return function (root, result) {
```
`result` may be unnecessary

gucong3000 replied:
It is ues `result` for result.stylelint.ignored
https://github.com/morishitter/stylefmt/pull/231/files#diff-f5008a4d411fbb1dbee72fa9d191a862R98

morishitter commented on lib/params.js:
```diff
@@ -52,10 +52,20 @@ function getIndentationFromStylelintRules (rules) {
-  return stylelint.getConfigForFile(file, configFile).then(function (stylelintrc){
-    return stylelintrc && stylelintrc.config && stylelintrc.config.rules
+  return Promise.resolve().then(function (){
+    if (file && !stylelint._options.ignoreDisables) {
```
in my opinion, we don't have to use the internal option of stylelint. what do you think about it?

gucong3000 commented:
## CHANGELOG
- Added: `--stdin-filename` option to CLI. same as [stylelint CLI](https://github.com/stylelint/stylelint/blob/master/src/cli.js#L75)
- Added: `--ignore-path` option to CLI. same as [stylelint CLI](https://github.com/stylelint/stylelint/blob/master/src/cli.js#L63)
- Added: `ignorePath` option to Node.js API. same as [stylelint API](http://stylelint.io/user-guide/node-api/#ignorepath)
- Added: `ignoreFiles` option to configuration file. same as [stylelint configuration](http://stylelint.io/user-guide/configuration/#ignorefiles)
- Added: `--ignore-disables` option to CLI.
- Added: `ignoreDisables` option to Node.js API.

morishitter commented on lib/params.js:
```diff
@@ -51,11 +51,21 @@ function getIndentationFromStylelintRules (rules) {
+    return stylelint.getConfigForFile(file, options.configFile).then(function (stylelintrc) {
+      return stylelintrc && stylelintrc.config && stylelintrc.config.rules
+    })
   }).catch(function (err) {
```
i'd like to use `catch()` in the bottom of this method chain (after here https://github.com/morishitter/stylefmt/pull/231/files#diff-f5008a4d411fbb1dbee72fa9d191a862R78 )

morishitter commented on bin/cli.js:
```diff
@@ -14,13 +14,15 @@ var argv = minimist(process.argv.slice(2), {
     d: 'diff',
+    h: 'help',
+    i: 'ignore-path',
+    id: 'ignore-disables',
```
please remove from here https://github.com/morishitter/stylefmt/pull/231#discussion_r88190351

gucong3000 replied:
ok

gucong3000 commented:
@morishitter 
All done.

morishitter commented:
@gucong3000 thanks! good stuff :)
